types/react-query/index.d.ts

```diff
@@ -36,107 +53,204 @@ export interface QueryOptions<TResult> {
+    status: 'success';
+    data: TResult[];
+    error: null;
+}
+
+export interface InfiniteQueryBaseState<TVariables> {
+    isFetching: boolean;
+    isFetchingMore: boolean;
+    failureCount: number;
+    refetch: (arg?: {
+        variables?: TVariables;
+        merge?: (...args: any[]) => any;
+        disableThrow?: boolean;
+    }) => Promise<any>;
+    fetchMore: (variables?: TVariables) => Promise<any>;
+    canFetchMore: boolean;
 }
 
 export function useMutation<TResults, TVariables extends object>(
     mutationFn: MutationFunction<TResults, TVariables>,
-    mutationOptions?: MutationOptions
-): [MutateFunction<TResults, TVariables>, MutationResult<TResults>];
+    mutationOptions?: MutationOptions<TResults>,
+): [MutateFunction<TResults, TVariables>, MutationState<TResults>];
 
-export type MutationFunction<TResults, TVariables extends object> = (
-    variables: TVariables,
-) => Promise<TResults>;
+export type MutationFunction<TResults, TVariables extends object> = (variables: TVariables) => Promise<TResults>;
 
-export interface MutationOptions {
-    refetchQueries?: Array<string | [string, object]>;
-    refetchQueriesOnFailure?: boolean;
+export interface MutationOptions<TResult> {
+    onSuccess?: (data: TResult) => Promise<any> | undefined;
+    onSettled?: (data: TResult, error: any) => Promise<any> | undefined;
+    onError?: (error: any) => Promise<any> | undefined;
+    throwOnError?: boolean;
+    useErrorBoundary?: boolean;
 }
 
 export type MutateFunction<TResults, TVariables extends object> = (
     variables?: TVariables,
     options?: {
-        updateQuery?: string | [string, object],
+        updateQuery?: string | [string, object];
         waitForRefetchQueries?: boolean;
-    }
+    },
 ) => Promise<TResults>;
 
-export interface MutationResult<TResults> {
-    data: TResults | null;
-    isLoading: boolean;
+export interface MutationState<TResults> {
+    status: 'idle' | 'loading' | 'error' | 'success';
+    data: TResults | undefined;
     error: null | Error;
     promise: Promise<TResults>;
-    reset: () => void;
 }
 
-export function setQueryData(
-    queryKey: string | [string, object],
-    data: any,
-    options?: {
-        shouldRefetch?: boolean
-    }
-): void | Promise<void>;
-
-export function refetchQuery(
-    queryKey: string | [string, object] | [string, false],
-    force?: {
-        force?: boolean
-    }
-): Promise<void>;
-
-export function refetchAllQueries(
-    options?: {
-        force?: boolean,
-        includeInactive: boolean
-    }
-): Promise<void>;
+export namespace queryCache {
+    function prefetchQuery<TResult, TQueryKey extends QueryKey>(
+        queryKey: TQueryKey,
+        queryFn: (key: string, ...args: QueryKeyVariablesArgs<TQueryKey>) => Promise<TResult>,
+        options?: QueryOptions<TResult>,
+    ): Promise<TResult>;
+
+    // tslint:disable:no-unnecessary-generics
+    function getQueryData<TResult>(queryKey: QueryKey): TResult | undefined;
```

Comment:
This type of generic definition is unsafe as there is nothing that guards the return value to ensure it is a `TResult` (which is why there is a lint error on this line). It might be better to return `unknown` here and have the caller of the function perform an appropriate cast or type guard.

Reply:
Isn't letting the caller perform the cast even more error-prone in this case? Even when they cast correctly to type associated with this query key, they may still forget to check for undefined value. Compare:

```
type IData = { foo: boolean};
const data = getQueryData<IData>('key'); // data is IData or undefined
data.foo // TS2532: Object is possibly 'undefined'.
```

vs
 
```
const data = getQueryData('key') as IData; 
data.foo // no error
```

Reply:
Changed the code followign your suggestion, but still explanation is much appreciated :-)